src/xwayland/x11_sockets.rs

```diff
@@ -0,0 +1,126 @@
+        }
+    }
+    // If we reach here, all values from 0 to 32 failed
+    // we need to stop trying at some point
+    return Err(());
+}
+
+/// Remove the X11 sockets for a given display number
+pub(crate) fn cleanup_x11_sockets(display: u32) {}
+
+pub(crate) struct X11Lock {
+    display: u32,
+}
+
+impl X11Lock {
+    /// Try to grab a lockfile for given X display number
+    fn grab(display: u32) -> Result<X11Lock, ()> {
+        let filename = format!("/tmp/.X{}-lock", display);
+        let lockfile = ::std::fs::OpenOptions::new()
+            .write(true)
+            .create_new(true)
+            .open(&filename);
+        match lockfile {
+            Ok(mut file) => {
+                // we got it, write our PID in it and we're good
+                let ret = file.write_fmt(format_args!("{:>10}", ::nix::unistd::Pid::this()));
+                if let Err(_) = ret {
+                    // write to the file failed ? we abandon
+                    ::std::mem::drop(file);
+                    let _ = ::std::fs::remove_file(&filename);
+                    return Err(());
+                } else {
+                    // we got the lockfile and wrote our pid to it, all is good
+                    return Ok(X11Lock { display });
+                }
+            }
+            Err(_) => {
+                // we could not open the file, now we try to read it
+                // and if it contains the pid of a process that no longer
+                // exist (so if a previous x server claimed it and did not
+                // exit gracefully and remove it), we claim it
+                // if we can't open it, give up
+                let mut file = ::std::fs::File::open(&filename).map_err(|_| ())?;
+                let mut spid = [0u8; 11];
+                file.read_exact(&mut spid).map_err(|_| ())?;
+                ::std::mem::drop(file);
+                let pid = ::nix::unistd::Pid::from_raw(::std::str::from_utf8(&spid)
+                    .map_err(|_| ())?
+                    .trim()
+                    .parse::<i32>()
+                    .map_err(|_| ())?);
+                if let Err(NixError::Sys(Errno::ESRCH)) = ::nix::sys::signal::kill(pid, None) {
+                    // no process whose pid equals the contents of the lockfile exists
+                    // remove the lockfile and try grabbing it again
+                    let _ = ::std::fs::remove_file(filename);
+                    return X11Lock::grab(display);
+                }
+                // if we reach here, this lockfile exists and is probably in use, give up
+                return Err(());
+            }
+        }
+    }
+
+    pub(crate) fn display(&self) -> u32 {
+        self.display
+    }
+}
+
+impl Drop for X11Lock {
+    fn drop(&mut self) {
+        // Cleanup all the X11 files
+        let _ = ::std::fs::remove_file(format!("/tmp/.X11-unix/X{}", self.display));
+        let _ = ::std::fs::remove_file(format!("/tmp/.X{}-lock", self.display));
```

Comment:
At least print a warning with the Err if it fails. `if let Err(err) = ... { warn!(this.logger, "Err deleting file: {}", err); }` or something.

Reply:
Yes, this is right, will add extensive logging.